### PR TITLE
Enable @FormParam for EntityParts

### LINKED ISSUE #25017: Posting Form-Data with the new Jakarta EE 10 Multipart Support fails
**Describe the bug**  
When posting multipart/form-data to a REST endpoint using the **@FormParam** annotation for an **EntityPart** or **InputStream** parameter, the request fails with a **400 Bad Request** response, and the following exception is logged:

```
jakarta.ws.rs.BadRequestException: RESTEASY003320: Failed processing arguments of public java.lang.String com.demo.rest.TestResource.upload(java.lang.String,jakarta.ws.rs.core.EntityPart) throws java.io.IOException
at org.jboss.resteasy.core.MethodInjectorImpl.injectArguments(MethodInjectorImpl.java:120)
Caused by: java.lang.UnsupportedOperationException: SRVE8020E: Servlet does not accept multipart requests
at com.ibm.ws.webcontainer.srt.SRTServletRequest.prepareMultipart(SRTServletRequest.java:3838)
```

During deployment, when using an **EntityPart** parameter, the following warning is logged:
`SROAP04005: Could not find schema class in index: jakarta.ws.rs.core.EntityPart`

**Steps to Reproduce**  
Using a simple REST endpoint like the following, causes the error:
    
```
@POST
@Consumes(MediaType.MULTIPART_FORM_DATA)
public Response upload(@FormParam("file") EntityPart file) {
    return Response.accepted().build();
}
```
**Expected behavior**  
The form-data is submitted successfully and in this example, a **202 Accepted** response is sent.

**Diagnostic information:**  
 - OpenLiberty Version: 23.0.0.3/wlp-1.0.75.cl230320230319-1900
 - Affected feature(s): jakartaee-10.0
 - Java Version: 
```
openjdk 17.0.6 2023-01-17
IBM Semeru Runtime Open Edition 17.0.6.0 (build 17.0.6+10)
Eclipse OpenJ9 VM 17.0.6.0 (build openj9-0.36.0, JRE 17 Linux amd64-64-Bit Compressed References 20230117_397 (JIT enabled,, AOT enabled)
OpenJ9   - e68fb241f
OMR      - f491bbf6f
JCL      - 927b34f84c8 based on jdk-17.0.6+10)
```
 - server.xml configuration: The default configuration from the current quickstarter:
```
<server description="test server">
  <featureManager>
    <feature>jakartaee-10.0</feature>
    <feature>microProfile-6.0</feature>
  </featureManager>
  <basicRegistry id="basic" realm="BasicRealm">
  </basicRegistry>
  <httpEndpoint id="defaultHttpEndpoint" httpPort="9080" httpsPort="9443" />
  <applicationManager autoExpand="true"/>
  <webApplication contextRoot="/test" location="test.war" />
  <ssl id="defaultSSLConfig" trustDefaultCerts="true" />
</server>

```

**Additional context**  
When using a **List&lt;EntityPart&gt; parts** method signature for the REST endpoint the error doesn't occur.

The problem can be mitigated by using a **web.xml** deployment descriptor with a suitable multipart configuration for the REST application, e.g.:

```
<servlet>
  <servlet-name>com.demo.rest.RestApplication</servlet-name>
  <multipart-config>
    <location>/tmp</location>
    <max-file-size>10485760</max-file-size>
    <max-request-size>2097152</max-request-size>
    <file-size-threshold>1048576</file-size-threshold>
  </multipart-config>
</servlet>
<servlet-mapping>
  <servlet-name>com.demo.rest.RestApplication</servlet-name>
  <url-pattern>/api/*</url-pattern>
</servlet-mapping>
```

### PR DESCRIPTION
**Pull request by jim-krueger**
Fixes #25017

### CODE REVIEW

**Comment by jim-krueger** [comment]
#libby #build

**Comment by jim-krueger** [comment]
#libby #build

**Comment by jim-krueger** [comment]
#build

**Comment by jim-krueger** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps

**Comment by jim-krueger** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps

**Comment by jim-krueger** [comment]
#libby

**Comment by jim-krueger** [comment]
#libby

**Comment by jim-krueger** [comment]
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

File: dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java [hunk 1]
```diff
@@ -68,6 +73,8 @@ public FormParamInjector(final Class type, final Type genericType, final Accessi
    @Override
    public Object inject(HttpRequest request, HttpResponse response, boolean unwrapAsync)
    {
+       boolean debug = tc.isDebugEnabled() && TraceComponent.isAnyTracingEnabled();  //Liberty Change
```
Comment (jhanders34) [inline dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java]:
```suggestion
       boolean debug = TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled();  //Liberty Change
```

File: dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java [hunk 2]
```diff
@@ -68,6 +73,8 @@ public FormParamInjector(final Class type, final Type genericType, final Accessi
...
+       List<String> list = null;
```
Comment (jhanders34) [inline dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java]: This should be lower after the EntityPart, List<EntityPart> and InputStream checks.

File: dev/io.openliberty.org.jboss.resteasy.server.ee10/src/io/openliberty/org/jboss/resteasy/server/component/RESTfulServletContainerInitializer.java [hunk 3]
```diff
@@ -17,6 +17,16 @@
 import java.util.Map;
 import java.util.Set;
 
+//import org.jboss.resteasy.plugins.server.servlet.HttpServlet30Dispatcher;
```
Comment (jhanders34) [inline dev/io.openliberty.org.jboss.resteasy.server.ee10/src/io/openliberty/org/jboss/resteasy/server/component/RESTfulServletContainerInitializer.java]:
```suggestion
```

File: dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartResource.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2022 IBM Corporation and others.
+ * Copyright (c) 2023 IBM Corporation and others.
```
Comment (jhanders34) [inline dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartResource.java]:
```suggestion
 * Copyright (c) 2022, 2023 IBM Corporation and others.
```

**Comment by jim-krueger** [comment]
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

**Review by jhanders34 — CHANGES_REQUESTED** [review]

**Comment by jhanders34** [comment]
You also should add a FormParam test with a mix of EntityPart and Strings.  Do you need more `@PUT` and `@GET` tests as well?

File: dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java (hunk 1 above)
Comment (jim-krueger) [reply dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java]: Done

File: dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java (hunk 2 above)
Comment (jim-krueger) [reply dev/io.openliberty.org.jboss.resteasy.common.ee10/src/org/jboss/resteasy/core/FormParamInjector.java]: Done

File: dev/io.openliberty.org.jboss.resteasy.server.ee10/src/io/openliberty/org/jboss/resteasy/server/component/RESTfulServletContainerInitializer.java (hunk 3 above)
Comment (jim-krueger) [reply dev/io.openliberty.org.jboss.resteasy.server.ee10/src/io/openliberty/org/jboss/resteasy/server/component/RESTfulServletContainerInitializer.java]: Done.

File: dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartResource.java (hunk 4 above)
Comment (jim-krueger) [reply dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartResource.java]: Done.

**Comment by jim-krueger** [comment]
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

**Review by WhiteCat22 — DISMISSED** [review]
I didn't walk through all the tests, but the product changes make sense to me

**Comment by jim-krueger** [comment]
#libby #build #spawn.fullfat.buckets=com.ibm.ws.jaxrs.2.0_fat,com.ibm.ws.jaxrs.2.0.cdi.1.2_fat,com.ibm.ws.jaxrs.2.0.client_fat,com.ibm.ws.jaxrs.2.1_fat,com.ibm.ws.jaxrs.2.1_fat_executor,com.ibm.ws.jaxrs.2.1_fat_extended,com.ibm.ws.jaxrs.2.1.cdi.2.0_fat,com.ibm.ws.jaxrs.2.1.client_fat,com.ibm.ws.jaxrs.2.1.sse_fat,com.ibm.ws.jaxrs.2.x_fat_clientProps,io.openliberty.restfulWS.3.0_fat,io.openliberty.restfulWS.3.0.cdi.3.0_fat,io.openliberty.restfulWS.3.0.client_fat,io.openliberty.restfulWS.3.1_fat

**Comment by jim-krueger** [comment]
Added logic to FormParamInjector to account for @FormParam multipart parameters on resource methods (specifically EntityPart, String, and InputStream per the specification).

In addition:  For some aspects of multipart processing, RESTEasy appears to require the @MultipartConfig annotation on a servlet.   For Liberty that translates to adding that annotation to our IBMRestServlet.   This ended up causing some NPEs in beanvalidation FAT until I added default values (replacing null) to config in WebAppConfiguratorHelper.   The addition of the annotation may be something we will want to remove at some point in the future.

File: dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartTestServlet.java [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2022 IBM Corporation and others.
+ * Copyright (c) 2023 IBM Corporation and others.
```
Comment (jhanders34) [inline dev/io.openliberty.restfulWS.3.1_fat/fat/src/io/openliberty/restfulWS31/fat/multipart/MultipartTestServlet.java]:
```suggestion
 * Copyright (c) 2022, 2023 IBM Corporation and others.
```

**Review by jhanders34 — APPROVED** [review]

**Comment by jim-krueger** [comment]
#libby

**Review by pmd1nh — APPROVED** [review]

**Comment by jim-krueger** [comment]
#run-libby-bot

**Pull request merged**